### ISSUE TITLE
lws_callback_on_writable() doesn't generate LWS_CALLBACK_HTTP_WRITEABLE on POST w/ no body

### ISSUE THREAD

**Issue by pblemel** [warmcat/libwebsockets#1421]
I am writing a server protocol that needs to support a POST without a body (a legacy client expects to be able to do this).  I get both LWS_CALLBACK_HTTP_BODY and LWS_CALLBACK_HTTP_BODY_COMPLETION callbacks (and return 0 for completeness).  In BODY_COMPLETION I write the headers and call lws_callback_on_writable() but never receive LWS_CALLBACK_HTTP_WRITEABLE.

If I use a test app (e.g. postman) and send a dummy body, everything works and WRITABLE is fired.  If using the same test app I omit the body (as the legacy client does) then WRITEABLE is never fired.

Respectfully,
Peter

**Comment by lws-team** [comment]
Zero-length POST bodies are supported, but you have to tell content-length: 0.

How can I reproduce your problem?

**Comment by pblemel** [comment]
Create a POST version of minimal-http-server-dynamic.c  Add the following lines below the path sprintf :  
```
                lws_snprintf(pss->path, sizeof(pss->path), "%s", (const char *)in);
                return 0;

        case LWS_CALLBACK_HTTP_BODY:
                return 0;

        case LWS_CALLBACK_HTTP_BODY_COMPLETION:
                /* prepare and write http headers... with regards to content-
                 * length, there are three approaches:
                 ...
```
From there I'm using the postman chrome plugin and issue a POST with no data.

Thanks for taking a look.
Peter

**Comment by lws-team** [comment]
Can you possibly send a patch on the minimal test app so I don't have to fiddle around?

**Comment by pblemel** [comment]
Yes. I'll create it from scratch from that server example and test it, just to be sure I didn't do something else to break it :)

**Comment by pblemel** [comment]
Well, creating it from scratch  ... works ...  :-/  There's something about the code when its in my app (practically line for line) that doesn't work when there's no body.  I don't use the same server startup incantation as the minimal app tho. 

The long and the short of it is that I need to do some more troubleshooting on my end and see if it's related to a particular configuration parameter (or if I added some other bug).

Thanks :)

**Comment by lws-team** [comment]
One hint about that, if you are calling through to the dummy callback by default, which is not a bad idea, take care that in the cases you handled something, you `return 0` out afterwards and don't just `break;` your way to doing the default as well.

**Comment by pblemel** [comment]
Something very strange is going on. I don't understand it, and it's going to take me a while to chase it down.

FWIW I am building and debugging this on a Linux Ubuntu 18 box (not my QNX host).  I downloaded and installed your patch last night.  Up until a few minutes ago, I got just the username from WSI_TOKEN_HTTP_AUTHORIZATION and never got the LWS_CALLBACK_HTTP_WRITEABLE callback.  I added a couple of debug writes using lwsl_notice (and nothing else, AFAIK), recompiled, and now  WSI_TOKEN_HTTP_AUTHORIZATION contains "Basic dXNlcjpwYXNzd29yZA==" (no longer just the username) and I am getting the LWS_CALLBACK_HTTP_WRITEABLE  callback.

I need to remove the lines I added and see if the behavior comes back.  Most peculiar.

**Comment by lws-team** [comment]
If your debug lines are not coming, check which .so your app is binding to with ldd...

**Comment by pblemel** [comment]
Okay... It's been a long day :).  

The good news is that my code now seems to be reliably getting LWS_CALLBACK_HTTP_WRITEABLE, with and without the debug writes.  The bad news is I have no idea what else I changed to make it work.  

Whatever I did, it's working and you can close this issue with my apologies.

At least I understand the WSI_TOKEN_HTTP_AUTHORIZATION mystery. I had a hypothesis that authentication might have something to do with the problem (your example doesn't use it, my code does, with few other differences), and had changed the auth file field to NULL in lws_http_mount to test the hypothesis.  LWS didn't need to authenticate the mount and just copied the authentication header into WSI_TOKEN_HTTP_AUTHORIZATION.  I'm fine with that, I had just forgotten I had made that change and was puzzled.  Putting authentication back takes care of that.

Thanks again,
Peter

**Comment by pblemel** [comment]
Okay!  I knew that I wasn't crazy (and didn't change something arbitrary). The problem is if you have debugger breakpoints in your code.  I have breakpoints in LWS_CALLBACK_HTTP and LWS_CALLBACK_HTTP_BODY_COMPLETION. If disable them (or 'continue' fairly quickly), I get the LWS_CALLBACK_HTTP_WRITEABLE callback.  If I take the time to inspect a variable or two, or am just slow 'continue'ing, I don't get the callback. 

This is why things seemed to suddenly work when I added debug writes.  I had disabled my breakpoints! 
 I was beginning to doubt my sanity.

I assume that there's a timer in your code somewhere that expects my callback to respond more quickly?

Thanks again,
Peter
